DM-45514: technote describing migration from `afw` types to new ShoeFits replacements (#1)
lsst-dm/dmtn-251: index.rst

   They make our on-disk file formats hard to document or read with external libraries, they're brittle and hard to maintain and evolve (largely because they force extensions to be written in C++).
   They rely on ``cfitsio``, which cannot perform partial reads on object stores, which will be absolutely critical in our hybrid-model Data Facility architecture.
   The types themselves suffer from numerous small problems that are hard to fix individually, such as the ``Mask`` plane dictionary singleton and the PSF "trampoline" system's reliance on poorly-supported ``pybind11`` functionality (leading in practice to a major memory leak).
   This technote will describe an approach to replacing these types that initially focuses on the serialization system and on-disk data models, while setting the stage for eventual deprecation of the ``Exposure`` and related types in the future.

Motivation
==========

The ``lsst.afw.image.Exposure`` class is currently used for almost all LSST image data products, from single-visit processed images to coadds to difference images.

> In addition to the main ``image`` plane, it holds an integer ``mask`` plane of the same size (where each bit has a different interpretation) and a floating point ``variance`` plane that records the per-pixel uncertainty (assuming it is uncorrelated).

Reviewer: ```suggestion
In addition to the main ``image`` plane, it holds an integer ``mask`` plane of the same size (where each bit corresponds to a particular mask field, e.g. `SATURATED`) and a floating point ``variance`` plane that records the per-pixel uncertainty (assuming it is uncorrelated).
```

There is currently no plan for implementing an alternate reader for old files - it's probably possible, but it would be easier to migrate any old processing runs we need to keep via a conversion script than support the old format indefinitely.
Keeping the old code around for a long time is not ideal, of course, but we will still reap substantial benefits from relying on it less, well before it is actually removed.

The ShoeFits File Format
========================

The way the new ShoeFits library maps Python objects to FITS files is heavily inspired by the `Advanced Scientific Data Format (ASDF) <https://asdf-standard.readthedocs.io/en/1.1.1/>`__, and especially its ASDF-in-FITS extension.
ASDF combines a YAML tree with a sequence of binary blocks, allowing most of a tree of complex objects to be represented in YAML, while allowing large arrays of floating point data to be pulled out of that tree stored naturally in binary form.
This is of course the same overall approach as FITS, but with FITS's archaic, limited header replaced by a modern hierarchical text format.
ASDF also defines a data model for common astrophysical primitives like celestial coordinates via JSON schema.

> ASDF-in-FITS is a little-used extension to the ASDF embeds the YAML tree and optional binary blocks into a FITS HDU as 1-d character image or binary table column.

Reviewer: ```suggestion
ASDF-in-FITS is a little-used extension to the ASDF that embeds the YAML tree and optional binary blocks into a FITS HDU as a 1-d character image or binary table column.
```

Even more importantly, library support for JSON in Python is in great shape, especially via the `Pydantic <https://docs.pydantic.dev/latest/>`__ library that ShoeFits builds on, and the same cannot be said for YAML.
The ShoeFits approach to data modeling is sufficiently similar to ASDF that it would be straightforward (but real work) to extend the ShoeFits library to support writing and reading ASDF files, without having to change any downstream ShoeFits-based serialization implementations.
ShoeFits explicitly adopts ASDF's JSON schemas for common astrophysical objects while using Pydantic to generate JSON schemas for custom objects, and those schemas provide the extra "tag" information that would be needed to turn a JSON tree (or an equivalent Python nested `dict`) into YAML.

ShoeFits nevertheless has a lot of FITS-specific interfaces because its goal is to map Python objects to *fully featured* FITS files, with FITS-standard WCSs and other headers that allow them to be at least mostly interpretable by third-party FITS readers that ignore the JSON tree.

> These FITS header cards are *exported* from the Python object tree: they are never read when reconstructing a Python object hierarchy from the serialized form, and hence we expect to duplicate the header information in the JSON tree (where it can often be expressed more naturally anyway).

Reviewer: Does this mean that the JSON tree will still be restricted to FITS-header representable things, e.g., <80chars, upper case keys, no NaN/inf (which have their own problems in JSON!), or can the FITS header export be not round-trippable?

> This helps keep the ShoeFits library simple: it only needs to be able to read the files it writes, not arbitrary external FITS files, and for this it only needs to be able to read JSON (which it can delegate to Pydantic) and FITS binary data.

Reviewer: I feel like `and FITS binary data` is carrying a lot of weight here. If you mean "as opposed to headers", that doesn't seem like a huge win, given the various forms of "binary table", "binary image", and compressed versions of all the above. If you really just mean "FITS binary image data", then that seems like a much bigger advantage.

If we add ASDF read/write support to ShoeFits in the future, the annotations used to declare FITS header exports would simply be ignored, and could be left out of any objects that would only be serialized to ASDF.

> Objects serializable with ShoeFits are also by definition serializable as plain JSON, though in some cases this would be a very bad idea (e.g. a 4k x 4k image would have to be be represented as a nested JSON array or base64-encoded byte string).

Reviewer: This seems to imply that we *wouldn't* want to serialize many of our objects with plain JSON? Is this paragraph really worth keeping? For which objects would this ability really be worth considering?

    import pydantic
    import lsst.shoefits as shf

    class Example(pydantic.BaseModel):
        array: shf.Array
        time: shf.Time
        instrument: str

At runtime these fields correspond to `numpy.ndarray` and `astropy.time.Time` (and `str`), respectively, and static type checkers like MyPy will see them that way as well.
When serialized to a JSON tree, this struct will use ASDF data models for representing them in JSON, and in the case of `~lsst.shoefits.Array`, the array data itself may be pulled out of the tree with a string pointer to a FITS HDU written in its place.
For this to work, the serialization needs to go through `~lsst.shoefits.FitsWriteContext` and `~lsst.shoefits.FitsReadContext`, which invoke Pydantic's JSON write/read logic with special hooks that intercept these annotations.

To control how fields appear in the FITS representation, parameterized annotations can be used::

    import pydantic
    import lsst.shoefits as shf
    from typing import Annotated

    class Example(pydantic.BaseModel):
        array: Annotated[shf.Array, shf.FitsOptions(extname="DATA")]
        time: shf.Time

>         instrument: Annotated[str, shf.ExportFitsHeaderKey("INSTRUME")]

Reviewer: Implementation question: `FitsOptions` and `ExportFitsHeaderKey` aren't semantically close for me. Does it make sense to group all of the "how this gets save in FITS" options in a similar pseudo-namespace? For example, changing these to `FitsOptions`, `FitsExportHeaderKey` would help them be more readily discoverable and tie their "FITS-ness" together.

> These FITS-specific annotations would be ignored by any other `~lsst.shoefits.WriteContext` or `~lsst.shoefits.ReadContext` implementation, but for the FITS implementations they result in the ``array`` field being writtend to an image extension HDU with ``EXTNAME='DATA'``, and the ``instrument`` field's value written to a FITS header ``INSTRUME`` key.

Reviewer: ```suggestion
These FITS-specific annotations would be ignored by any other `~lsst.shoefits.WriteContext` or `~lsst.shoefits.ReadContext` implementation, but for the FITS implementations they result in the ``array`` field being written to an image extension HDU with ``EXTNAME='DATA'``, and the ``instrument`` field's value written to a FITS header ``INSTRUME`` key.
```

Serialization by Proxy
----------------------

In order to make a custom type serializable via a separate model class, the mapping can be defined via an `~lsst.shoefits.Adapter` subclass::

    import pydantic
    import lsst.shoefits as shf
    from typing import Annotated, TypeAlias

    class Thing:
        """An arbitrary type that isn't serializable directly."""

        def __init__(self, a: int):
            self._a = a

    class ThingModel(pydantic.BaseModel):
        """A model type representing the serializable form of `Thing`."""
        a: int

    class ThingAdapter(shf.Adapter[Thing, ThingModel]):
        """Class that declares serialization for `Thing` via `ThingModel`."""

        @property
        def model_type -> type[ThingModel]:
            return ThingModel

        def to_model(self, thing: Thing) -> ThingModel:
            return ThingModel(a=thing._a)

        def from_model(self, model: ThingModel) -> Thing:
            return Thing(model.a)

    SerializableThing: TypeAlias = Annotated[Thing, ThingAdapter]

> Any parent model type can now declare a field of type ``SerializableThing``, which will be a ``Thing`` instance at runtime and in static type checkers, while being transparently serializable via ``ThingModel``.

Reviewer: Have you made a list of how many `Adapter` classes we'll need to serialize all of the things we currently write to our image files? I'm curious whether there would be a lot of these to implement, and whether any of them would be particularly complicated.

ShoeFits Primitives
-------------------

ShoeFits defines a few new serializable classes that can be used as building blocks for higher-level models:

> - `~lsst.shoefits.Box` is analogous to `lsst.geom.Box2I`, but is not limited to two dimensions.

Reviewer: Why only `Box2I`, not `Box2D`? I find myself using the latter almost as much as the former.

  This is a Pydantic model and is hence directly serializable.
- `~lsst.shoefits.Interval` is analogous to `lsst.geom.IntervalI`, and exists largely to help implement `~lsst.shoefits.Box`.
  It is also a Pydantic model.
- `~lsst.shoefits.Image` is analogous to `lsst.afw.image.Image`: it combines 2-d`numpy.ndarray` with a 2-d integer offset that can be used to match a subimage's coordinate system to that of its parent image.

>   It also has an optional `astropy.units.Unit`, allowing images with units to be saved via the ASDF ``Quantity`` data model.

Reviewer: Woo! `njy` images here we come!

Though, why make the unit optional?

  `~lsst.shoefits.Image` is not a Pydantic model, but it implements Pydantic's special custom-object serialization hooks in the same way the `~lsst.shoefits.Array` type alias does, allowing it to be stored in-line in the JSON tree or (usually) in a separate FITS image HDU.
  When saved to FITS HDU, a special FITS WCS is written to represent the integer offset coordinate system, just as with `lsst.afw.image.Image`.
- `~lsst.shoefits.Mask` is analogous to `lsst.afw.image.Mask`, i.e. an integer bitmask, but it delegates management of the mask planes to a separate `~lsst.shoefits.MaskSchema` object that can be shared by `~lsst.shoefits.Mask` instances without being a singleton.

>   Unlike its ``afw`` counterpart, `~lsst.shoefits.Mask` is backed by a 3-d array, with the last dimension used to support dynamic number of mask planes: typically the ``dtype`` of a `~lsst.shoefits.Mask` backing array is just ``unit8``, and the shape of the last dimension is the number of mask planes divided by eight.

Reviewer: That dynamic length might be useful, but I wonder if it will cause extra headaches in both implementation and user problems. There are still uses for "get the bitmask and do bit-wise math on the Mask", and I'm not sure we could come up with a rich enough API for Mask to fully prevent that. I'm also not sure it's necessary, since we're just barely hitting the limits of 32-bit uint masks right now, and I suspect 64-bits would be more than enough. And mapping between a dynamic shoefits Mask and the afw fixed-bits one seems painful. Maybe there's some other context I'm missing here?

        def to_model(self, box: Box2I) -> shf.Box:
            return shf.Box.factory[box.getSlices()]
        def from_model(self, model: shf.Box) -> Box2I:
            return Box2I(Point2I(box.x.min, box.x.max), Point2I(box.y.min, box.y.max))

Defining these adapters will be one of the first tasks involved in implementing the ShoeFits models to replace ``Exposure``.

These adapters can be used directly as an annotation, i.e. ``Annotated[Box2I, BoxAdapter]``, but in many cases it will be easier to just call the adapter methods directly in some higher-level `~lsst.shoefits.Adapter`.

Polymorphism
------------

Some of our most important ``Exposure`` component types are instances of an abstract base class or reliant internally on one, such as `lsst.afw.detection.Psf` or `lsst.afw.math.BoundedField`.
When the set of possible subclasses can be fully enumerated when a parent model type is defined, it's often best to do just enumerate them using `typing.Union` or the equivalent ``|`` syntax::

    import pydantic

    class A(pydantic.BaseModel):
        value: int

    class B(pydantic.BaseModel):
        value: str

    class Holder(pydantic.BaseModel):
        nested: A | B

> Pydantic's `discriminated union <https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions>`__ functionality can be used to make this pattern much more efficient and robus (the default implementation on read is to try all possibilities until one works).

Reviewer: ```suggestion
Pydantic's `discriminated union <https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions>`__ functionality can be used to make this pattern much more efficient and robust (the default implementation on read is to try all possibilities until one works).
```

    class Adapter1(shf.Adapter[Derived1, Model1]):

        @property
        def model_type(self) -> type[Model1]:
            return Model1

        def to_model(self, d: Derived1) -> Model1:
            return Model1()

        def from_model(self, m: Model1) -> Derived1:
            return Derived1()

And finally register it::

    registry = shf.PolymorphicAdapterRegistry()
    registry.register_adapted("one", Adapter1())

For the case where a polymorphic implementation class is itself a Pydantic model, we can skip the adapter class and register it as native::

    class Derived2(Base, pydantic.BaseModel):
        value: int

        def get_value(self) -> int:
            return self.value

        def _get_tag(self) -> str:
            return "two"

    registry.register_native("two", Derived2)

In both cases, the registration step needs to happen before reading begins, which means that it is not sufficient to declare a global `~lsst.shoefits.PolymorphicAdapterRegistry` instance in a low-level package and have downstream packages add their own types to it at import time, unless something else takes responsibility for importing them before read attempts are made.

> The exact mechanism we'll use to solve this problem is TBD; ShoeFits tries to be agnostic to it, largely to avoid imposing any solution that might run afoul of security concerns (e.g. importing modules according to serialized strings).

Reviewer: I'm a little worried about this. I know we've sometimes had trouble with the existing measurement registries when someone didn't import the right package, and requiring it before a read seems definitely worrying. Do you have a few options you're considering for the TBD mechanism?

I worry that the potential flexibility gained by being able to define storage types elsewhere will be outweighed by the complexity added in import ordering from just requiring that all types be defined in shoefits directly.

In all the above, the definition of the ``Base`` ABC is actually unimportant: it's generally good practice, of course, as a way to clearly define an interface, and static type checkers will see ``Annotated[Base, Polymorphic(...)]`` as just ``Base``, as usual with `Annotated`.

> But neither ShoeFits nor Pydantic will actually check that objects loaded via the adaper registry actually inherit from ``Base``, so it's perfectly viable for the annotation to be a `typing.Protocol` or even just `typing.Any` or `object`, if the interface isn't formalized or strongly typed.

Reviewer: I'm not sure about the purpose of this sentence. What interface wouldn't be strongly typed? I thought the entire point of this design was to use strong typing to build the interface automatically?

> In addition to the complexity involved in setting up an adapter registry, the major disadvantage of `~lsst.shoefits.Polymorphic` is that there is no way to emit JSON schema for polymorphic fields, since the parent model type has no information about how those fields will be serialized.

Reviewer: If we can't emit schema for those types, how do we actually use them in practice?

Reading Compressed Subimages
============================

> ShoeFits mostly sidesteps the limitations of ``cfitsio`` in terms of reading from object stores by delegating to `astropy.io.fits` instead.

Reviewer: Is `astropy.io.fits` a requirement here, or could one substitute the python-based `fitsio` instead? I know that there are still cases where astropy fits is markedly slower, but maybe this system can avoid those?

At first it would support only lossless GZIP compression, since that's what we actually use today, and it would not do any FITS header parsing; it'd expect callers to use `astropy.io.fits` for that in Python (as ShoeFits will).
It also won't do any I/O itself: it will return byte range addresses for callers to fetch, and then accept those raw bytes and transform them into image data.
It would be designed to grow into something usable as the backend for a more general-purpose FITS reader like Astropy's, even if its initial focus is extremely narrow.
Keeping it separate from ShoeFits will hopefully help it pick up external contributions in this direction, in addition to keeping the complexity of a compiled extension module out of ShoeFits itself.

I would like to implement this module in Rust, and have the DM stack consume it via conda (which makes it unnecessary for the stack or ``rubin-env`` to include any Rust tooling).
I believe the interfaces and scope are simple enough that we can make that work in this case without requiring more-frequent-than-usual updates to the ``rubin-env`` metapackage.

> Rust's libraries and toolchain for compiled Python extension modules will make it much easier (relative to C or C++) to set up the package and leverage third-party decompression code (for GZIP), without making it any even harder to call small bits of vendored CFITSIO code if that becomes useful.

Reviewer: ```suggestion
Rust's libraries and toolchain for compiled Python extension modules will make it much easier (relative to C or C++) to set up the package and leverage third-party decompression code (for GZIP), without making it any harder to call small bits of vendored CFITSIO code if that becomes useful.
```

> Rust's libraries and toolchain for compiled Python extension modules will make it much easier (relative to C or C++) to set up the package and leverage third-party decompression code (for GZIP), without making it any even harder to call small bits of vendored CFITSIO code if that becomes useful.

Reviewer: "ll small bits of vendored CFITSIO code if that becomes useful." - Now I'm sweating. I think we have to decide whether that is something we want to support in any way, or whether we want to fully avoid it. Though maybe you're thinking of small and specific enough examples that it would be mostly harmless?

"We built this tiny Rust library for handling in-memory compressed FITS blocks, but you also need cfitsio for these other three things, though we do all the important stuff in astropy.io.fits" doesn't seem like a good look.

While Rust would be a new language for DM-maintained code, we have a few developers with Rust experience already, and I believe the benefits outweigh the drawbacks.

The Long-Term Trajectory of the LSST Science Pipelines
======================================================

By proposing the replacement and eventual retirement of many of our workhorse `lsst.afw` classes, this proposal is implicitly suggesting a very different future for much of our codebase, beyond the changes it directly proposes.
And while I don't want the relatively concrete ShoeFits proposal to get tied up a long debate about competing visions, my own (still vague) vision for the future of the stack did play a role in putting this proposal together (and wanting to use Rust in part of it).

So, with the caveat that this section is *context* for the RFC, not part of the RFC itself:

> - I'd like each separate git package to have its own semver releases, with dependencies between packages handled by standard packaging tools like ``pip`` and ``conda``.

Reviewer: I think I like this part of the plan. It's already almost true for geom, sphgeom, and afw, given how slowly they evolve, so it feels like a reasonable ask.

  We may then be able to start refactoring these into packages that are genuinely lightly coupled (or at least only coupled in the right dependency direction), but we have to start by having our package structure reflect the real code dependencies that exist.

- I'd like to get rid of all of our cross-package C++ interfaces, and with them our standalone C++ shared libraries and unit tests.
  These shared libraries are *the* major impediment to distributing our code via PyPI.
  They're also a big piece of why the stack is slow to build and import, though it's not clear to what extent this is intrinsic to having a lot of shared libraries or an interaction with other atypical packaging procedures.
  This mostly involves moving data structures up to Python and passing simpler types to C++ code.
  Merging packages also helps here, because it transforms cross-package C++ interfaces into intra-package C++ interfaces.

> - I'd like to write a new low-level astronomical geometry library that combines the functionality in `lsst.sphgeom`, `lsst.geom`, and `lsst.afw.geom`, while improving consistency and avoiding subtle but hard-to-fix mistakes in the current types.

Reviewer: [xkcd 927](http://xkcd.com/927/) rears its hydra-like head.

I hope this is very far down your schedule (or as far as it can be, given that it may be a requirement of a fully-working shoefits). Putting together our AST-based WCS in afw took Russell far longer than we initially expected. Maybe you'd be able to leverage much of that work so that this isn't as big of a project as it sounds like, but saying "we'll write a python/C API for astronomy primitives" (assuming your next paragraph, "move aware from C++" applies here) sounds like at least a year's worth of work, given past experience.

  This would include our wrappers for AST (which I'd like to abstract away a bit more) and all of the mid-level data structures I envision us needing to define in a compiled language for downstream compiled-langauge modules to reuse.
  That means that it should export a Python C API, to allow those downstream modules to build against it without needing a separate shared library.
  Leaving room for this future library (without assuming it will be available anytime soon) is one reason why the geometry types in ShoeFits are as simple as possible.

- I'd like to move away from C++ wherever we can in implementations as well as interfaces.
  This mostly means preferring Python, as we already do.

>   But I also think we should consider JIT-compilation tools like JAX and Numba as alternatives to C++ for algorithmic code, especially when something can be expressed via algorithmic primitives that may benefit from GPUs in the future.

Reviewer: I have not looked at those tools for our use cases in several years, but I think the danger of going that direction is similar to the danger of relying on C++: it's hard to find enough developers with experience. Just having python experience doesn't mean that you can just wrap regular python in Numba and get a free win.

I'm not sure I have a good alternative proposal here, but thinking about *who* we can get to work on our code is probably one of the more important considerations in choosing tooling.

  And we should consider Rust as an alternative to C++ in cases where a compiled language is needed but the problem involves complex data structures and/or systems programming rather than numerics.

>   I do expect that there will be cases where C++ will remain the best choice due to either history or the availability of certain third-party libraries, but I'd like to isolate and minimize these.

Reviewer: Same comment as above: is it easier or harder to find Rust developers in astronomy, as compared with C++ developers? How many Rust libraries are there that are as good as e.g. Eigen or GSL?
